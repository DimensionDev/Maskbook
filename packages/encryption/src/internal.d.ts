type Readwrite<T> = {
    -readonly [key in keyof T]: T[key]
}
