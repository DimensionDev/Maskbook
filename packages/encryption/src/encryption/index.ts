export * from './Decryption'
export * from './Encryption'
export * from './AppendEncryption'
