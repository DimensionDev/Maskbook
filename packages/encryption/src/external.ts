// This file is the entry for external usage.
export {
    encodePayload,
    parsePayload,
    EC_KeyCurveEnum,
    SocialNetworkEnum,
    SocialNetworkEnumToProfileDomain,
    type EC_Key,
    type Signature,
    type PayloadParseResult,
    type PayloadWellFormed,
    type SupportedPayloadVersions,
} from './payload/index.js'

export {
    decrypt,
    DecryptError,
    DecryptProgressKind,
    DecryptIntermediateProgressKind,
    DecryptErrorReasons,
    type DecryptOptions,
    type DecryptIO,
    type DecryptEphemeralECDH_PostKey,
    type DecryptStaticECDH_PostKey,
    type DecryptProgress,
    type DecryptIntermediateProgress,
    type DecryptReportedInfo,
    type DecryptSuccess,
} from './encryption/index.js'

export { socialNetworkEncoder, socialNetworkDecoder } from './social-network-encode-decode/index.js'

export { type TypedMessage, type Meta } from '../../typed-message/dist/base/index.js'
export * from '../../typed-message/dist/base/core/index.js'
export {
    ProfileIdentifier,
    PostIVIdentifier,
    CheckedError,
    OptionalResult,
    type AESCryptoKey,
    type EC_Public_CryptoKey,
    type EC_Private_CryptoKey,
    type EC_CryptoKey,
} from '@masknet/shared-base'

export * from 'ts-results-es'
