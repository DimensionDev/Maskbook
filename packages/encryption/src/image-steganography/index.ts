import { encode, decode, GrayscaleAlgorithm } from '@dimensiondev/stego-js'
import type { EncodeOptions } from '@dimensiondev/stego-js'
import { getDimension } from './utils.js'
import { getPreset, findPreset } from './presets.js'
import { encodeArrayBuffer } from '@dimensiondev/kit'

export { GrayscaleAlgorithm, AlgorithmVersion } from '@dimensiondev/stego-js'

export interface SteganographyIO {
    downloadImage: (url: string) => Promise<ArrayBuffer>
}
export interface EncodeImageOptions extends SteganographyIO {
    data: string | Uint8Array
    password: string
    grayscaleAlgorithm?: GrayscaleAlgorithm
    preset: SteganographyPreset
}
export enum SteganographyPreset {
    Preset2021 = '2021',
    Preset2022 = '2022',
}

export async function steganographyEncodeImage(buf: ArrayBuffer, options: EncodeImageOptions) {
    const { downloadImage, data, password, grayscaleAlgorithm } = options
    const preset = getPreset(options.preset)
    if (!preset) throw new Error('Failed to find preset.')

    const optionalOptions: Partial<Readwrite<EncodeOptions>> = {}
    if (grayscaleAlgorithm) optionalOptions.grayscaleAlgorithm = grayscaleAlgorithm

    const text = typeof data === 'string' ? data : encodeArrayBuffer(data)

    return new Uint8Array(
        await encode(buf, await downloadImage(preset.mask), {
            ...preset.options,
            ...optionalOptions,
            text,
            pass: password,
        }),
    )
}

export type DecodeImageOptions = SteganographyIO & Partial<EncodeOptions> & Pick<EncodeOptions, 'pass'>

export async function steganographyDecodeImage(image: Blob | string, options: DecodeImageOptions) {
    const buffer = typeof image === 'string' ? await options.downloadImage(image) : await image.arrayBuffer()
    const dimension = getDimension(buffer)
    const preset = findPreset(dimension)
    if (!preset) return ''
    return decode(buffer, await options.downloadImage(preset.mask), {
        ...preset.options,
        ...options,
    })
}
