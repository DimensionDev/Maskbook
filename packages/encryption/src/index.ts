export {
    encodePayload,
    parsePayload,
    EC_KeyCurveEnum,
    SocialNetworkEnum,
    SocialNetworkEnumToProfileDomain,
    type EC_Key,
    type Signature,
    type PayloadParseResult,
    type PayloadWellFormed,
    type SupportedPayloadVersions,
} from './payload'

export {
    encrypt,
    EncryptError,
    EncryptErrorReasons,
    type EncryptOptions,
    type EncryptIO,
    type EncryptResult,
    type EncryptTargetE2E,
    type EncryptionResultE2EMap,
    type EncryptTargetPublic,
    type EncryptionResultE2E,
} from './encryption'

export {
    decrypt,
    DecryptError,
    DecryptProgressKind,
    DecryptIntermediateProgressKind,
    DecryptErrorReasons,
    type DecryptOptions,
    type DecryptIO,
    type DecryptEphemeralECDH_PostKey,
    type DecryptStaticECDH_PostKey,
    type DecryptProgress,
    type DecryptIntermediateProgress,
    type DecryptReportedInfo,
    type DecryptSuccess,
} from './encryption'

export {
    //
    type AppendEncryptionIO,
    type AppendEncryptionOptions,
    appendEncryptionTarget,
} from './encryption'

export { socialNetworkEncoder, socialNetworkDecoder, TwitterDecoder } from './social-network-encode-decode'

export {
    type DecodeImageOptions,
    type EncodeImageOptions,
    type ImageTemplateTypes,
    type SteganographyIO,
    steganographyDecodeImage,
    steganographyEncodeImage,
    GrayscaleAlgorithm,
} from './image-steganography'

// TODO: remove them in the future
export { importEC_Key, getEcKeyCurve, importAES as importAESFromJWK } from './utils'
