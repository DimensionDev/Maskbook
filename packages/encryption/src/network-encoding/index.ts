import { EncryptPayloadNetwork } from '../payload/types.js'
import { sharedDecoder, sharedEncoder } from './shared.js'
import { TwitterDecoder, __TwitterEncoder } from './twitter.js'

export function decodeByNetwork(network: EncryptPayloadNetwork, content: string): Array<string | Uint8Array> {
    if (network === EncryptPayloadNetwork.Twitter) {
        return TwitterDecoder(content)
            .map((x) => [x])
            .unwrapOr([])
    }
    const possiblePayload = content.match(/(\u{1F3BC}[\w+/=|]+:\|\|)/giu) || []

    const result: Array<string | Uint8Array> = []
    for (const payload of possiblePayload) {
        const decoded = sharedDecoder(payload)
        if (decoded.isSome()) result.push(decoded.value)
        else result.push(payload)
    }
    return result
}
export function encodeByNetwork(network: EncryptPayloadNetwork, content: string | Uint8Array): string {
    // v38
    if (typeof content === 'string') {
        if (network === EncryptPayloadNetwork.Twitter) return __TwitterEncoder(content)
        return content
    }

    // v37
    if (network === EncryptPayloadNetwork.Twitter) return __TwitterEncoder(content)
    return sharedEncoder(content)
}

export * from './twitter.js'
