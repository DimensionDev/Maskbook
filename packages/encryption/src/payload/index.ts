import type { PayloadWellFormed } from '..'
import { encode37, encode38, parse37, parse38, parse39, parse40, PayloadParserResult } from '../payload_internal'
import { EKindsError as Err, EKinds, EKindsError } from '../types'

export * from './types'
export async function parsePayload(payload: unknown): PayloadParserResult {
    if (payload instanceof ArrayBuffer) {
        return parse37(payload)
    }
    if (typeof payload === 'string') {
        if (payload.startsWith('ðŸŽ¼4/4')) return parse38(payload)
        if (payload.startsWith('ðŸŽ¼3/4')) return parse39(payload)
        if (payload.startsWith('ðŸŽ¼2/4')) return parse40(payload)
    }
    return new Err(EKinds.InvalidPayload, 'Unknown version').toErr()
}

export function encodePayload(payload: PayloadWellFormed.Payload) {
    if (payload.version === -38) return encode38(payload)
    else if (payload.version === -37) return encode37(payload)
    return new EKindsError(EKinds.EncodeFailed, 'unsupported payload version').toErr()
}
