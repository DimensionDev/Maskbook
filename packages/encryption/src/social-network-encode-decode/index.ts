import { SocialNetworkEnum } from '../payload/types'
import { sharedDecoder, sharedEncoder } from './shared'
import { TwitterDecoder, __TwitterEncoder } from './twitter'

export function socialNetworkDecoder(network: SocialNetworkEnum, content: string): Array<string | Uint8Array> {
    if (network === SocialNetworkEnum.Twitter) {
        return TwitterDecoder(content)
            .map((x) => [x])
            .unwrapOr([])
    }
    const newVersion = sharedDecoder(content)
    if (newVersion.some) return [newVersion.val]

    const v38 = content.match(/(\u{1F3BC}[\w+/=|]+:\|\|)/giu)
    if (v38) return v38
    return []
}
export function socialNetworkEncoder(network: SocialNetworkEnum, content: string | Uint8Array): string {
    // v38
    if (typeof content === 'string') {
        if (network === SocialNetworkEnum.Twitter) return __TwitterEncoder(content)
        return content
    }

    // v37
    if (network === SocialNetworkEnum.Twitter) return __TwitterEncoder(content)
    return sharedEncoder(content)
}

export * from './twitter'
