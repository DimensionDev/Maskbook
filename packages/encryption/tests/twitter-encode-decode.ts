import { test, expect } from '@jest/globals'
import { TwitterDecoder } from '../src/social-network-encode-decode'
test('Can decode Twitter text payload', () => {
    const payload = `This tweet is encrypted with #mask_io (@realMaskNetwork). ðŸ“ªðŸ”‘

    Install https://mask.io/?PostData_v1=%204/4.mFXgbxPDrAD6PYRHWJW5IVlHObuFYs/LUyd08kkUkKimTx4opjHXjrgorrsyTPidL0lkeqdXnW3EzEacdlPJChko-Q5Wy113UMHYy1aV1Wheqwy/vM3PLMVaoKW/HyIL/SjJtim75bWMSHmj+vzMYwV59jLJqlTjJqhz5ObdZhAsWwRJGYIzEkg+.SKlWOvMpyF9T1jL2pr95kA_=.ehROC4d3QudYHEABcBTpSuxLvxM=._.Aq/bVWAKvodJuURGk3enjE1gUiu2SELM8IIKIlNGqOWM.1.dHdpdHRlci5jb20vd2lsbHlvbnRvdXI=%40 to decrypt it.`
    expect(TwitterDecoder(payload)).toMatchSnapshot('Twitter example')
})
