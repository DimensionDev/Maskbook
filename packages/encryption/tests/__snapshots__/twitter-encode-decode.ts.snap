// Vitest Snapshot v1

exports[`Can decode Twitter text payload > Twitter example 1`] = `
SomeImpl {
  "none": false,
  "some": true,
  "val": "ðŸŽ¼4/4|mFXgbxPDrAD6PYRHWJW5IVlHObuFYs/LUyd08kkUkKimTx4opjHXjrgorrsyTPidL0lkeqdXnW3EzEacdlPJChko+Q5Wy113UMHYy1aV1Wheqwy/vM3PLMVaoKW/HyIL/SjJtim75bWMSHmj+vzMYwV59jLJqlTjJqhz5ObdZhAsWwRJGYIzEkg+|SKlWOvMpyF9T1jL2pr95kA==|ehROC4d3QudYHEABcBTpSuxLvxM=|_|Aq/bVWAKvodJuURGk3enjE1gUiu2SELM8IIKIlNGqOWM|1|dHdpdHRlci5jb20vd2lsbHlvbnRvdXI=:||",
}
`;

exports[`Can decode Twitter text payload: Twitter example 1`] = `Some("ðŸŽ¼4/4|mFXgbxPDrAD6PYRHWJW5IVlHObuFYs/LUyd08kkUkKimTx4opjHXjrgorrsyTPidL0lkeqdXnW3EzEacdlPJChko+Q5Wy113UMHYy1aV1Wheqwy/vM3PLMVaoKW/HyIL/SjJtim75bWMSHmj+vzMYwV59jLJqlTjJqhz5ObdZhAsWwRJGYIzEkg+|SKlWOvMpyF9T1jL2pr95kA==|ehROC4d3QudYHEABcBTpSuxLvxM=|_|Aq/bVWAKvodJuURGk3enjE1gUiu2SELM8IIKIlNGqOWM|1|dHdpdHRlci5jb20vd2lsbHlvbnRvdXI=:||")`;
